lighthouse-core/lib/network-request.js

```diff
@@ -0,0 +1,260 @@
+    this.redirects = /** @type {NetworkRequest[]|undefined} */ (undefined);
+    this.failed = false;
+    this.localizedFailDescription = '';
+
+    this._initiator = /** @type {LH.Crdp.Network.Initiator} */ ({type: 'other'});
+    this._timing = /** @type {LH.Crdp.Network.ResourceTiming|undefined} */ (undefined);
+    this._resourceType = /** @type {LH.WebInspector.ResourceType|undefined} */ (undefined);
+    this._mimeType = '';
+    this.priority = () => /** @type {LH.Crdp.Network.ResourcePriority} */ ('Low');
+    this.initiatorRequest = () => /** @type {NetworkRequest|undefined} */ (undefined);
+    this._responseHeaders = /** @type {LH.WebInspector.HeaderValue[]} */ ([]);
+
+    this._fetchedViaServiceWorker = false;
+    this._frameId = /** @type {string|undefined} */ ('');
+    this._isLinkPreload = false;
+
+    // Make sure we're compatible with old WebInspector.NetworkRequest
+    // eslint-disable-next-line no-unused-vars
+    const record = /** @type {LH.WebInspector.NetworkRequest} */ (this);
+  }
+
+  /**
+   * @return {NetworkRequest}
+   */
+  clone() {
+    return Object.assign(new NetworkRequest(), this);
+  }
+
+  /**
+   * @param {NetworkRequest} initiator
+   */
+  setInitiatorRequest(initiator) {
+    this.initiatorRequest = () => initiator;
+  }
+
+  /**
+   * @param {LH.Crdp.Network.RequestWillBeSentEvent} data
+   */
+  onRequestWillBeSent(data) {
+    this.requestId = data.requestId;
+    this._requestId = data.requestId;
+
+    const url = new URL(data.request.url);
+    this.url = data.request.url;
+    this._url = data.request.url;
+    this.documentURL = data.documentURL;
+    this.parsedURL = {
+      scheme: url.protocol.split(':')[0],
+      // Intentional, DevTools uses different terminology
+      host: url.hostname,
+      securityOrigin: () => url.origin,
+    };
+    this.isSecure = SECURE_SCHEMES.includes(this.parsedURL.scheme);
+
+    this.startTime = data.timestamp;
+
+    this.requestMethod = data.request.method;
+
+    this._initiator = data.initiator;
+    this._resourceType = data.type && resourceTypes[data.type];
+    this.priority = () => data.request.initialPriority;
+
+    this._frameId = data.frameId;
+    this._isLinkPreload = data.initiator.type === 'preload' || !!data.request.isLinkPreload;
+  }
+
+  onRequestServedFromCache() {
+    this._fromMemoryCache = true;
+  }
+
+  /**
+   * @param {LH.Crdp.Network.ResponseReceivedEvent} data
+   */
+  onResponseReceived(data) {
+    this._onResponse(data.response, data.timestamp, data.type);
+    this._frameId = data.frameId;
+  }
+
+  /**
+   * @param {LH.Crdp.Network.DataReceivedEvent} data
+   */
+  onDataReceived(data) {
+    this._resourceSize += data.dataLength;
+    if (data.encodedDataLength !== -1) {
+      this.transferSize += data.encodedDataLength;
+    }
+  }
+
+  /**
+   * @param {LH.Crdp.Network.LoadingFinishedEvent} data
+   */
+  onLoadingFinished(data) {
+    if (this.finished) return;
```

Comment:
when does this happen?

Reply:
it seems to be expected for a certain class of request failures (aborted redirected requests)

not sure how to best comment it...maybe?
`// On cancelled requests, DevTools can send duplicate erroneous events, prefer the first one for best timing data`

Reply:
I think that's 👍